File: src/vec2.c
#ifndef __AG_VEC2__
#define __AG_VEC2__

#include <math.h>

#include "tools.h"

typedef struct Vec2 {
  double x, y;
} Vec2;

Vec2 ag_vec2_origin(void);
Vec2 ag_vec2_random(const double max_x, const double max_y);
double ag_vec2_length(const Vec2 *vector);
void ag_vec2_normalise(Vec2 *vector);
Vec2 ag_vec2_normalised(Vec2 vector);

Vec2 ag_vec2_origin(void) { return (Vec2){.x = 0, .y = 0}; }

Vec2 ag_vec2_random(const double max_x, const double max_y) {
  return (Vec2){
      .x = double_random(max_x),
      .y = double_random(max_y),
  };
}

double ag_vec2_length(const Vec2 *vector) {
  return sqrt(pow(vector->x, 2) + pow(vector->y, 2));
}

void ag_vec2_normalise(Vec2 *vector) {
  double length = ag_vec2_length(vector);
  vector->x /= length;
  vector->y /= length;
}

Vec2 ag_vec2_normalised(Vec2 vector) {
  ag_vec2_normalise(&vector);
  return vector;
}

#endif // __AG_VEC2__

File: src/vec2.zig
const std = @import("std");
const expectEqual = std.testing.expectEqual;

pub const tools = @import("tools.zig");

pub const Vec2 = struct {
    x: f64,
    y: f64,

    pub fn origin() @This() {
        return .{ .x = 0, .y = 0 };
    }

    pub fn random(max_x: f64, max_y: f64) @This() {
        return .{
            .x = tools.double_random(max_x),
            .y = tools.double_random(max_y),
        };
    }

    pub fn length(self: Vec2) f64 {
        return @sqrt((self.x * self.x) + (self.y * self.y));
    }

    pub fn normalise(self: *Vec2) void {
        const len = self.length();
        self.x /= len;
        self.y /= len;
    }

    pub fn normalised(self: Vec2) Vec2 {
        var v = self;
        v.normalise();
        return v;
    }
};

test "origin" {
    const v = Vec2.origin();
    try expectEqual(v.x, 0);
    try expectEqual(v.y, 0);
}

test "length" {
    const v1 = Vec2{ .x = 1, .y = 0 };
    try expectEqual(v1.length(), 1.0);
    const v2 = Vec2{ .x = 0, .y = 1 };
    try expectEqual(v2.length(), 1.0);
    const v3 = Vec2{ .x = 1, .y = 1 };
    try expectEqual(v3.length(), @sqrt(2.0));
    const v4 = Vec2{ .x = 3, .y = 4 };
    try expectEqual(v4.length(), 5.0);
}

test "normalised" {
    const v = Vec2.random(100, 100).normalised();
    try expectEqual(v.length(), 1.0);
}
